pkg/aws/eni/limits/limits.go

```diff
@@ -867,19 +114,79 @@ func UpdateFromEC2API(ctx context.Context, api ec2API) error {
 	return nil
 }
 
-// parseLimitString returns the Limits struct parsed from config string.
-func parseLimitString(limitString string) (limit ipamTypes.Limits, err error) {
-	intSlice := make([]int, 3)
-	stringSlice := strings.Split(strings.ReplaceAll(limitString, " ", ""), ",")
-	if len(stringSlice) != 3 {
-		return limit, fmt.Errorf("invalid limit value")
-	}
-	for i, s := range stringSlice {
-		intLimit, err := strconv.Atoi(s)
+func (l *LimitsGetter) newLimitsUpdateTrigger(api ec2API, minInterval, timeout time.Duration, retryCount int) (*trigger.Trigger, error) {
+	return trigger.NewTrigger(trigger.Parameters{
+		Name:        "ec2-instance-limits-update",
+		MinInterval: minInterval,
+		TriggerFunc: func(reasons []string) {
+			if err := l.updateFromEC2APIWithRetry(context.TODO(), api, timeout, retryCount); err != nil {
+				log.WithError(err).Warning("Failed to update instance limits from EC2 API")
+			}
+
+			l.Lock()
+			l.lastUpdate = time.Now()
+			select {
+			case l.triggerDone <- struct{}{}:
+			default:
+				log.Debug("Skipping trigger done signal as channel is full")
+			}
+			l.Unlock()
+		},
+	})
+}
+
+func (l *LimitsGetter) InitEC2APIUpdateTrigger(api ec2API, triggerMinInterval, ec2APITimeout time.Duration, ec2APIRetryCount int) error {
+	var initErr error
+	l.initOnce.Do(func() {
+		l.Lock()
+		l.triggerMinInterval = triggerMinInterval
+		l.ec2APITimeout = ec2APITimeout
+		l.ec2APIRetryCount = ec2APIRetryCount
+		l.Unlock()
+
+		t, err := l.newLimitsUpdateTrigger(api, triggerMinInterval, ec2APITimeout, ec2APIRetryCount)
 		if err != nil {
-			return limit, err
+			log.WithError(err).Warning("Failed to create EC2 update trigger")
+			initErr = err
```

Comment:
Not that error handling in a `sync.Once` is dangerous - if this case here happens, the trigger will never ever be created again (since `sync.Once` will never execute the body again, regardless of what happened here). Now in this case here it's fine, since `NewTrigger` actually cannot fail and the line highlighted here is dead code, but I think it would be nicer to remove the usage of `sync.Once` if possible here (see sibling comment).

For a more in-depth example of why `if` inside `sync.Once` is dangerous, see https://victoriametrics.com/blog/go-sync-once/